public/static/settings.js

```diff
@@ -0,0 +1,83 @@
+// This function handles the toggling of selections of all upstream search engines 
+// options in the settings page under the tab engines.
+function toggleAllSelection() {
+  document
+    .querySelectorAll('.engine')
+    .forEach(
+      (engine_checkbox) =>
+        (engine_checkbox.checked =
+          document.querySelector('.select_all').checked)
+    )
+}
+
+// This function adds the functionality to sidebar buttons to only show settings 
+// related to that tab.
+function setActiveTab(current_tab) {
+  document
+    .querySelectorAll('.tab')
+    .forEach((tab) => tab.classList.remove('active'))
+  document
+    .querySelectorAll('.btn')
+    .forEach((tab) => tab.classList.remove('active'))
+  current_tab.classList.add('active')
+  document
+    .querySelector(`.${current_tab.innerText.toLowerCase().replace(' ', '_')}`)
+    .classList.add('active')
+}
```

Comment:
The second function, setActiveTab, uses querySelectorAll which can be slow if there are many elements to search through. It also uses forEach which can be slow if there are many elements to iterate through. A more performant solution would be to use a for loop instead of forEach.
```suggestion
function setActiveTab(current_tab) {
  const tabs = document.querySelectorAll('.tab');
  const buttons = document.querySelectorAll('.btn');
  for (let i = 0; i < tabs.length; i++) {
    tabs[i].classList.remove('active');
  }
  for (let i = 0; i < buttons.length; i++) {
    buttons[i].classList.remove('active');
  }
  current_tab.classList.add('active');
  const tabName = current_tab.innerText.toLowerCase().replace(' ', '_');
  document.querySelector(`.${tabName}`).classList.add('active');
}
```